src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -1007,7 +1003,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     }
 
     /// Gets a package for the given package id.
-    pub fn get_package(&self, id: &PackageId) -> CargoResult<&'a Package> {
+    pub fn get_packages(&self, id: &[&PackageId]) -> CargoResult<Vec<&'a Package>> {
```

Comment:
So this method is pretty interesting. Long ago Cargo had the signature:

```rust
pub fn get_package(&self, id: &PackageId) -> &'a Package;
```

but we then changed to [lazily download packages](https://github.com/rust-lang/cargo/pull/2406) in order to close https://github.com/rust-lang/cargo/issues/2394. That was the easiest fix to do at the time but I think is causing issues with this patch.

I think to get the most benefit out of the scheme (downloading in parallel) we'll want to probably implement a separate pass over `Context` to collect a list of packages that we're gonna want to download. Perhaps that could be deferred to a future PR, though? (just want to make sure we don't lose track of this)

Reply:
I'm willing to do some refactor, thanks for the reference!

Reply:
FWIW I've been thinking recently that the end-state here may be to "future-ify" everything where we just create a gob of futures and throw them at tokio to do the whole build, although that's certainly not a refactoring required for this PR!